### PR TITLE
Pools: Add ability to configure commission claiming permissions

### PR DESCRIPTION
**Pull request by rossbulat**
Addresses #409.

This request has been raised by multiple community members - the ability for the nomination pool root role to configure permissionless commission claiming:

> Would it be possible to have a claim_commission_other extrinsic for claiming commission of nomination pools permissionless?

This PR does not quite introduce this additional call, but amends `do_claim_commission` to check a new `claim_permission` field in the `Commission` struct, configured by an enum:

```
enum CommissionClaimPermission {
   Permissionless,
   Account(AccountId),
}
```
This can be optionally set in a bonded pool's `commission.claim_permission` field: 

```
struct BondedPool {
   commission: {
      <snip>
      claim_permission: Option<CommissionClaimPermission<T::AccountId>>,
   },
   <snip>
}
```

This is a new field and requires a migration to add it to existing pools. This will be `None` on pool creation, falling back to the `root` role having sole access to claim commission if it is not set; this is the behaviour as it is today. Once set, the field _can_ be set to `None` again.

#### Changes
- [x] Add `commision.claim_permission` field.
- [x] Add `can_claim_commission` and amend `do_claim_commission`.
- [x] Add `set_commission_claim_permission` call.
- [x] Test to cover new configs and call.
- [x] Add and amend benchmarks.
- [x] Generate new weights + slot into call `set_commission_claim_permission`.
- [x] Add migration to introduce `commission.claim_permission`, bump storage version.
- [x] Update Westend weights.
- [x] Migration working.

### CODE REVIEW

**Comment by rossbulat** [comment]
bot bench substrate-pallet --pallet=pallet_nomination_pools

**Comment by rossbulat** [comment]
bot bench substrate-pallet --pallet=pallet_nomination_pools

**Comment by rossbulat** [comment]
bot bench substrate-pallet --pallet=pallet_nomination_pools

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4461098

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4461240

**Comment by rossbulat** [comment]
bot bench polkadot-pallet --pallet=pallet_nomination_pools --runtime=westend

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4468226

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4468366

**Comment by rossbulat** [comment]
bot bench polkadot-pallet --pallet=pallet_nomination_pools --runtime=westend

**Comment by rossbulat** [comment]
bot bench polkadot-pallet --pallet=pallet_nomination_pools --runtime=westend

File: substrate/frame/nomination-pools/src/migration.rs [hunk 1]
```diff
@@ -47,6 +57,101 @@ pub mod versioned {
...
+			let onchain = Pallet::<T>::on_chain_storage_version();
+
+			log!(
+				info,
+				"Running migration with current storage version {:?} / onchain {:?}",
+				current,
+				onchain
+			);
+
+			if onchain == 7 {
```
Comment (gpestana) [inline substrate/frame/nomination-pools/src/migration.rs:112]: I don't think we need to manually check the version of the storage and current migration with the `VersionedMigration` and the code looks cleaner and less error prone imo. Maybe we should use that pattern instead?

A good example [here](https://github.com/paritytech/polkadot-sdk/blob/cfa19c37e60f094a2954ffa73b1da261c050f61f/substrate/frame/grandpa/src/migrations/v5.rs)

**Review by gpestana — CHANGES_REQUESTED** [review]

File: substrate/frame/nomination-pools/src/lib.rs [hunk 2]
```diff
@@ -1078,6 +1088,17 @@ impl<T: Config> BondedPool<T> {
...
+	fn can_claim_commission(&self, who: &T::AccountId) -> bool {
+		if let Some(permission) = self.commission.claim_permission.as_ref() {
+			match permission {
+				CommissionClaimPermission::Permissionless => true,
+				CommissionClaimPermission::Account(account) => account == who || self.is_root(who),
+			}
+		} else {
+			self.is_root(who)
+		}
+	}
```
Comment (Ank4n) [inline substrate/frame/nomination-pools/src/lib.rs:1100]: nit or probably just preference but I would check `self.is_root` first and break early if that's the case.

File: substrate/frame/nomination-pools/src/migration.rs [hunk 3]
```diff
@@ -47,6 +57,82 @@ pub mod versioned {
...
+	impl<T: Config> OnRuntimeUpgrade for VersionUncheckedMigrateV7ToV8<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let mut translated = 0u64;
+			BondedPools::<T>::translate::<OldBondedPoolInner<T>, _>(|_key, old_value| {
+				translated.saturating_inc();
+				Some(old_value.migrate_to_v8())
+			});
+
+			StorageVersion::new(8).put::<Pallet<T>>();
+			log!(info, "Upgraded {} pools, storage to version 8", translated);
```
Comment (Ank4n) [inline substrate/frame/nomination-pools/src/migration.rs:109]: Storage version is already set by the outer migration code. Also a generic message is logged.
Refer https://github.com/paritytech/polkadot-sdk/blob/2b7f73c651c419953dea49947096129d3782c9b4/substrate/frame/support/src/migrations.rs#L118

**Review by Ank4n — APPROVED** [review]
Some small nits to be fixed, looks good overall.

File: substrate/frame/nomination-pools/src/lib.rs [hunk 4]
```diff
@@ -2742,6 +2768,28 @@ pub mod pallet {
 			let who = ensure_signed(origin)?;
 			Self::do_adjust_pool_deposit(who, pool_id)
 		}
+
```
Comment (gpestana) [inline substrate/frame/nomination-pools/src/lib.rs:2771]: Maybe add rustdoc to the callable? I find them very useful in the interfaces (e.g. polkadotjs).

**Review by gpestana — APPROVED** [review]

File: substrate/frame/nomination-pools/src/lib.rs (hunk 4 above)
Comment (rossbulat) [reply substrate/frame/nomination-pools/src/lib.rs:2771]: Good spot, an oversight on my part. Thanks!

**Pull request merged**